explauto/interest_model/riac.py

```diff
@@ -0,0 +1,947 @@
+        
+        """
+        if sampling_mode is None:
+            sampling_mode = self.sampling_mode
+            
+        #print sampling_mode
+        
+        if sampling_mode[0] == 'random':
+            rand_mode = 'by_volume'
+            if rand_mode == 'by_leaf': 
+                # Choose a leaf randomly
+                return np.random.choice(self.get_leaves()).sample_bounds()
+                
+            elif rand_mode == 'by_volume': 
+                # Choose a leaf weighted by volume, randomly
+                if self.leafnode:
+                    return self.sample_bounds()
+                else:
+                    split_ratio = (self.split_value - self.bounds_x[0,self.split_dim]) / (self.bounds_x[1,self.split_dim] - self.bounds_x[0,self.split_dim])
+                    if split_ratio > np.random.random():
+                        return self.less.sample(sampling_mode=['random'])
+                    else:
+                        return self.greater.sample(sampling_mode=['random'])
+                
+        elif sampling_mode[0] == 'greedy':
+            # Choose the leaf with the max progress
+            if self.leafnode:
+                return self.sample_bounds()
+            else:
+                lp = self.less.progress
+                gp = self.greater.progress
+                if gp > lp:
+                    return self.greater.sample(sampling_mode=['greedy'])
+                else:
+                    return self.less.sample(sampling_mode=['greedy'])
+                
+        elif sampling_mode[0] == 'epsilon_greedy':
+            # Choose the leaf with the max progress with probability (1-eps) and a random leaf with probability (eps).
+            eps = sampling_mode[1]
+            if eps > np.random.random():
+                return self.sample(sampling_mode=['random'])
+            else:
+                return self.sample(sampling_mode=['greedy'])
+                
+        elif sampling_mode[0] == 'softmax':
+            # Sample leaves with probabilities progress*volume and a softmax exploration (with a temperature parameter)
+            if self.leafnode:
+                return self.sample_bounds()
+            else:
+                temperature = sampling_mode[1]
+                
+                leaves = self.get_leaves()
+                #print leaves[0].leafnode
+                #print map(lambda leaf:leaf.progress, leaves)
+                progresses = np.array(map(lambda leaf:leaf.progress*leaf.volume(), leaves)) #by volume
+                progress_max = max(progresses)
+                probas = np.exp(progresses / (progress_max*temperature))
+                probas = probas / np.sum(probas)
+                
+                #print "Softmax", progresses, probas
+                leaf = leaves[np.where(np.random.multinomial(1, probas) == 1)[0][0]]
+                return leaf.sample_bounds()
+            
+        else:
+            raise NotImplementedError(sampling_mode)
```

Comment:
This method is actually quite hard to follow. Maybe a solution would be to split them into methods for each sampling mode. And then the sample method could just be something like:
```python

def sample(sampling_mode=None):
    return self.sample_methods[samplig_mode]()
```